### PR TITLE
Test more cases of Java Toolchain usage

### PR DESCRIPTION
**Pull request by alllex**
This is a follow-up for #21281 

This PR adds more integration tests for the JVM toolchain usage events. The tests cover the cases when java home is directly set on the `CompileJava` task, executable is set on the `Test` task, and toolchains are configured for the Kotlin plugin.

### CODE REVIEW

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (wolfs) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:77]: Good call about removing the `stripIndent()`.

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (wolfs) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:145]: When this PR is in, we should extend the test to also get toolchain progress events under the `ResolveTaskMutationsBuildOperationType` operations. So this would work as well for up-to-date tasks.

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (wolfs) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:105]: This should work after merging https://github.com/gradle/gradle/pull/21363, right?

**Comment by alllex** [comment]
@bot-gradle test this

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (wolfs) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:396]: So Kotlin works, interesting! I suppose it would fail when running this test with configuration caching enabled.

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (wolfs) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:300]: Can we point to an issue here?

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (wolfs) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:275]: I think it would be interesting to run with having one toolchain for the project (`runWithInstallation`) and another one for the `javaHome`. Would it then emit two different toolchain events?

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (wolfs) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:401]: How is test related here?

**Review by wolfs — COMMENTED** [review]
Nice tests! I left some comments and some questions.

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (alllex) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:145]: I think so as well. I'll wait until #21363 is merged and then try updating the tests for this

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (alllex) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:105]: Not sure, because IIRC it does not work due to `Javadoc` task not having a `JavaVersion` input, like `JavaCompile` and `Test` do.

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (alllex) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:401]: It checks that even though we only configure the toolchain within the `kotlin` block, it actually affects the java launcher selected by the `test` task.

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (alllex) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:353]: I am not sure it is the best solution to keep Kotlin version hardcoded here.
However, I wasn't able to quickly figure out if we have some variable that hold something like latest supported Kotlin plugin version or similar.

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (alllex) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:300]: Done

**Comment by alllex** [comment]
@bot-gradle test qfl

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (alllex) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:432]: Yes, it fails. I excluded the test run with the configuration cache and linked a new issue for this.

**Comment by alllex** [comment]
@bot-gradle test qfl

**Comment by alllex** [comment]
@wolfs I tried enabling the test blocks that test for up-to-dateness, but none of the tests started passing. So I am keeping them as is, just merged the master.

**Comment by alllex** [comment]
@bot-gradle test this

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (wolfs) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:324]: That is interesting. Why does `jdkMetadata1` not show up?

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (wolfs) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:366]: I think it would be good to assert that there are no events here, so the test fails when we fix this.

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (wolfs) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:390]: We have a class `KotlinGradlePluginVersions` which you can use in a where clause to test multiple Kotlin versions.

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (wolfs) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:617]:
```suggestion
        def taskRecord = operations.first(buildOperationType) {
```

I suppose the number can go away again.

**Review by wolfs — COMMENTED** [review]
This looks pretty good! I left some more comments.

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (alllex) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:324]: I presume it is because of the way the toolchain selection is implemented in the `JavaCompile`:
https://github.com/gradle/gradle/blob/81fcbf4528111618cb2116f0d530f6957b030656/subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java#L236-L263

It may also be due to the fact that `JavaCompile` (along with `Test` and `Javadoc`) does not set a convention on the tool property.

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (alllex) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy]: Done

**Comment by alllex** [comment]
Updated the tests, addressing review comments

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (wolfs) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:37]: It would be good to test non-stable releases as well, so we can give them feedback before the final release is out.

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (wolfs) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:327]: Right, though there still is https://github.com/gradle/gradle/blob/b81eac42a17b7249ee7d2fe76e81e0326e0b1f94/subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java#L123-L133

I suppose if you configure a toolchain in the `java` project block, then two will show up. Can you add/modify a test for this?

**Review by wolfs — COMMENTED** [review]
Nearly there.

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (wolfs) [inline subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:449]: Would it make sense to test both tasks in one go?

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (alllex) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy]: Done

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (alllex) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy]: Done

**Review by wolfs — APPROVED** [review]
LGTM!

File: subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy
Comment (alllex) [reply subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy:327]: If additionally a toolchain is configured via the `java` block, then a validation kicks in that fails the `compileJava` task.
```
java {
    toolchain {
        languageVersion = JavaLanguageVersion.of(${jdkMetadata1.languageVersion.majorVersion})
    }
}
```

```
Execution failed for task ':compileJava'.
> Must not use `javaHome` property on `ForkOptions` together with `javaCompiler` property
```

**Comment by alllex** [comment]
@bot-gradle test and merge

**Comment by alllex** [comment]
@bot-gradle test and merge

**Comment by alllex** [comment]
@bot-gradle test this

**Comment by alllex** [comment]
@bot-gradle test and merge

**Pull request merged**
